feat: add caching around json hasher

scasplte2 opened the pull request.

scasplte2 committed:
feat: add caching around json hasher

scasplte2 commented:
a rough comparison between the cached and uncached hashers using the worksheet [HasherPerformanceTest.sc](https://github.com/user-attachments/files/22522081/HasherPerformanceTest.txt)

This shows a two order of magnitude improvement in even moderately sized objects (around 100 kB) 

```
============================================================
Hasher Performance Comparison Test
============================================================

📊 Testing with ~0.1MB object
----------------------------------------
📈 Cached Hasher:
   • Average: 0.09 ms
   • Min: 0.03 ms | Max: 0.61 ms
   • Total time: 9.08 ms
   • Cache hit rate: ~99.0%

📉 Uncached Hasher:
   • Average: 13.18 ms
   • Min: 11.98 ms | Max: 22.25 ms
   • Total time: 1317.94 ms

🚀 Performance Improvement:
   • Speedup: 151.32x faster with cache
   • Time saved per hash: 13.09 ms
   • Total time saved: 1308.86 ms

✅ Hash consistency check: PASSED

📊 Testing with ~0.5MB object
----------------------------------------
📈 Cached Hasher:
   • Average: 0.10 ms
   • Min: 0.04 ms | Max: 0.20 ms
   • Total time: 9.96 ms
   • Cache hit rate: ~99.0%

📉 Uncached Hasher:
   • Average: 68.69 ms
   • Min: 62.14 ms | Max: 96.91 ms
   • Total time: 6869.38 ms

🚀 Performance Improvement:
   • Speedup: 694.15x faster with cache
   • Time saved per hash: 68.59 ms
   • Total time saved: 6859.42 ms

✅ Hash consistency check: PASSED

📊 Testing with ~1.0MB object
----------------------------------------
📈 Cached Hasher:
   • Average: 0.09 ms
   • Min: 0.04 ms | Max: 0.31 ms
   • Total time: 9.43 ms
   • Cache hit rate: ~99.0%

📉 Uncached Hasher:
   • Average: 136.25 ms
   • Min: 126.11 ms | Max: 146.81 ms
   • Total time: 13625.59 ms

🚀 Performance Improvement:
   • Speedup: 1457.06x faster with cache
   • Time saved per hash: 136.16 ms
   • Total time saved: 13616.16 ms

✅ Hash consistency check: PASSED

📊 Testing with ~2.0MB object
----------------------------------------
📈 Cached Hasher:
   • Average: 0.09 ms
   • Min: 0.04 ms | Max: 0.22 ms
   • Total time: 8.71 ms
   • Cache hit rate: ~99.0%

📉 Uncached Hasher:
   • Average: 273.71 ms
   • Min: 257.20 ms | Max: 302.03 ms
   • Total time: 27371.29 ms

🚀 Performance Improvement:
   • Speedup: 3167.63x faster with cache
   • Time saved per hash: 273.62 ms
   • Total time saved: 27362.57 ms

✅ Hash consistency check: PASSED

📊 Testing with ~5.0MB object
----------------------------------------
📈 Cached Hasher:
   • Average: 0.09 ms
   • Min: 0.04 ms | Max: 0.21 ms
   • Total time: 9.24 ms
   • Cache hit rate: ~99.0%

📉 Uncached Hasher:
   • Average: 702.38 ms
   • Min: 657.12 ms | Max: 776.25 ms
   • Total time: 70238.82 ms

🚀 Performance Improvement:
   • Speedup: 7655.49x faster with cache
   • Time saved per hash: 702.29 ms
   • Total time saved: 70229.57 ms

✅ Hash consistency check: PASSED

============================================================
Test Complete!
============================================================




🔄 SAME OBJECT TEST (Best case for cache)
============================================================

val sameObjectTest: LargeObject = LargeObject(e17aa370-9209-4a62-bfb2-4a661dbf2914,List(NestedData(value_c62528c1-c12d-422d-838a-ff2fdda0ea7c,List(0.21227356829532262, 0.37103873745820015, 0.3843806387041626, 0.39382591740769424, 0.7048404856890214, 0.4487563023771234, 0.6748387996124593, 0.09747689493728973, 0.10140695712731362, 0.6358781287363361),Some(SubData(field_c43f5be1-2289-41e7-ba30-33bd676efb10,5896,List(1fdf2c96-1473-44d3-8087-d02709da19e9, 19ca61bd-a1e1-4fe8-ae80-001999da5f92, 06e7e723-2a8c-4a6c-bc1f-e269a2ef6278, 3714625c-96c8-49ae-8df1-2b705cce86fd, 1472234c-dff6-4737-bb3f-3d657db3549f)))), NestedData(value_e6469120-34a0-4500-9edb-e9f4c411180b,List(0.551200959318769, 0.6522588345445742, 0.8919602698896393, 0.010499483847732582, 0.8459887553811841, 0.6408064597267739, 0.262...
val cachedHasher2: io.constellationnetwork.security.Hasher[cats.effect.IO] = io.constellationnetwork.security.Hasher$$anon$4@56cb820c
val uncachedHasher2: io.constellationnetwork.security.Hasher[cats.effect.IO] = io.constellationnetwork.security.Hasher$$anon$5@4abba876


val sameObjIterations: Int = 50

val cachedSameStart: Long = 944096509217955



val cachedSameEnd: Long = 944096952737549

val uncachedSameStart: Long = 944097009923667



val uncachedSameEnd: Long = 944111039926708

val cachedSameTotal: Double = 443.519594
val uncachedSameTotal: Double = 14030.003041

Cached: 443.52 ms total (8.87 ms avg)
Uncached: 14030.00 ms total (280.60 ms avg)
Speedup: 31.63x faster with cache
```

scasplte2 committed:
feat: add feature flag around constructing hasher

scasplte2 committed:
fix: address PR comments

scasplte2 commented:
@greptileai update your review please

scasplte2 commented on modules/shared/src/main/scala/io/constellationnetwork/security/Hasher.scala:
```diff
@@ -88,7 +91,49 @@ object Hasher {
+    }
+
+  def forJsonCached[F[_]: Sync: JsonSerializer]: Hasher[F] = {
+    val cache: Cache[Any, Hash] = Scaffeine()
```
we may miss but the purpose of the cache is to avoid serialization and hashing so this will use hashCode and equality checking to determine if the cache is hit. This could be better but it should work for repeated objects.

scasplte2 commented on modules/shared/src/main/scala/io/constellationnetwork/security/Hasher.scala:
```diff
@@ -88,7 +91,49 @@ object Hasher {
+      .recordStats()
+      .expireAfterAccess(5.minutes)
+      .maximumSize(10000)
+      .softValues()
```
SoftValues seems preferred over soft keys

  What softKeys() would do (Greptile's suggestion):

  - Would keep the keys (the data objects) using soft references
  - This would be problematic because:
    - If the key gets garbage collected, you lose the cache entry
    - You can't look up a cache entry if the key has been GC'd
    - Makes the cache much less effective

  Why your current setup is correct:

  val cache: Cache[Any, Hash] = Scaffeine()
    .recordStats()
    .expireAfterAccess(5.minutes)
    .maximumSize(10000)
    .softValues()  // ← This is correct!
    .build[Any, Hash]()

  - Strong references to keys: The data objects you're hashing
  - Soft references to values: The computed hashes
  - This ensures you can always look up a cached hash if the object is still around
  - The hash values can be evicted under memory pressure and recomputed if needed

scasplte2 commented on modules/shared/src/main/scala/io/constellationnetwork/security/Hasher.scala:
```diff
@@ -88,7 +91,49 @@ object Hasher {
+        hashJson(data).map(_ === expectedHash)
+
+      def hashJson[A: Encoder](data: A): F[Hash] =
+        Sync[F].delay(cache.getIfPresent(data)).flatMap {
+          case Some(hash) => hash.pure[F]
+          case None =>
+            computeHash(data).flatTap { hash =>
+              Sync[F].delay(cache.put(data, hash))
+            }
+        }
```
defeats the purpose of this cache

ryle-goehausen commented on modules/shared/src/main/scala/io/constellationnetwork/security/Hasher.scala:
```diff
@@ -88,7 +91,49 @@ object Hasher {
 
-  def forJson[F[_]: Sync: JsonSerializer]: Hasher[F] = new Hasher[F] {
+  def forJson[F[_]: Sync: JsonSerializer]: Hasher[F] =
+    sys.env.get("CL_CACHE_JSON_HASH").exists(_.toLowerCase == "true") match {
```
I think default behavior should be to use this, we only need to disable it for testnet comparisons.

scasplte2 replied:
addressed

ryle-goehausen commented on modules/shared/src/main/scala/io/constellationnetwork/security/Hasher.scala:
```diff
@@ -88,7 +91,49 @@ object Hasher {
+    val cache: Cache[Any, Hash] = Scaffeine()
+      .recordStats()
+      .expireAfterAccess(5.minutes)
+      .maximumSize(10000)
```
Both of these should maybe be configurations, but not necessary, could be higher potentially but this is safe.

scasplte2 replied:
agreed but I'm not sure how most cleanly to do this (how to avoid touching all the invocations of `Hasher` creation). I considered a default value which may work but leaving for now to test

ryle-goehausen commented:
@scasplte2 bump, I think we should merge this soon as we can test on testnet with default true.

scasplte2 committed:
refactor: update flag name and use cached hasher by default

scasplte2 commented:
> @scasplte2 bump, I think we should merge this soon as we can test on testnet with default true.

apologies, I overlooked that I had pending comments and let this get stale. I've rebased onto `develop` and addressed using the cache by default
